File: 687.最长同值路径/longestUnivaluePath.go
package problem0687

type TreeNode struct {
	Val   int
	Left  *TreeNode
	Right *TreeNode
}

func longestUnivaluePath(root *TreeNode) int {
	if root == nil {
		return 0
	}
	var (
		result = 0
		dfs    func(node *TreeNode) int
	)
	// 考虑一个递归：搜寻以node为起点的最长同值路径
	dfs = func(node *TreeNode) int {
		if node == nil {
			return 0
		}
		l := dfs(node.Left)
		r := dfs(node.Right)
		left, right := 0, 0
		if node.Left != nil && node.Val == node.Left.Val {
			left = l + 1
		}
		if node.Right != nil && node.Val == node.Right.Val {
			right = r + 1
		}
		result = max(result, left+right)
		return max(left, right)
	}
	dfs(root)
	return result
}

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

File: 687.最长同值路径/longestUnivaluePath.js
// 这道题的递归还是细
function longestUnivaluePath(root) {
	if (!root) {
		return 0
	}
	let result = 0

	const dfs = node => {
		if (!node) {
			return 0
		}
		let left = 0
		let right = 0
		const l = dfs(node.left)
		const r = dfs(node.right)
		if (node.left && node.left.val === node.val) {
			left = l + 1
		}
		if (node.right && node.right.val === node.val) {
			right = r + 1
		}
		// 更新最大值时，考虑双侧
		result = Math.max(result, left + right)
		// dfs返回值返回的node节点是单侧的最长同值路径的长度
		return Math.max(left, right)
	}

	dfs(root)

	return result
}